### ISSUE TITLE
Agents don't follow their prompts 2

### ISSUE THREAD

**Issue by cth127** [google-deepmind/concordia#95]
I made the same title with #78, but a slightly different issue.

I'm testing with `gemme-2-9b-it` and an agent (including baseline `rational_agent`) does not follow its own chain of thougt.

![image](https://github.com/user-attachments/assets/12711587-37c7-4d3f-9821-b6e292e70c37)

As shown in the picture above, an agent thinks The Duke of York is the best option, however, selects The Crooked Billet.

There are too many cases that hinder the robust evaluation. (Promise me, nearly 1 out of 3 cases returns such an error)

I guess a better prompting rule is required.

**Comment by cth127** [comment]
Another example on haggling

![image](https://github.com/user-attachments/assets/102a6ce0-899f-4c5c-94f4-2c5d475920dd)

**Comment by cth127** [comment]
I guess this comes from the different `sample_choice` method on `together_ai.py` from the other APIs. Why does only `together_ai.py` utilize log probability-based `sample_choice`?

**Comment by kevin-v96** [comment]
I think I also faced this issue during the hackathon, while using Codestral.

**Comment by jzleibo** [comment]
There are many reasons they might get confused and respond inconsistently to statements earlier in their chain of thought. One possibility could be that they are getting confused in a kind of needle in a haystack way if they have a lot of text in context at the same time. Or perhaps there is another ambiguous but potentially contradictory statement somewhere else in the chain of thought. This kind of thing usually requires agent-specific and llm-specific debugging to get it to work well. The baselines we provided are certainly not the best they could be in this regard.

**Comment by vezhnick** [comment]
> I guess this comes from the different `sample_choice` method on `together_ai.py` from the other APIs. Why does only `together_ai.py` utilize log probability-based `sample_choice`?

Most models do not support logits as output, but together.ai does. Using logits is both faster and "cleaner", in a sense of genuinely picking the most likely answer. 
As Joel mentioned, there could be many reasons the agent is not strictly following it's chain-of-thought. It also depends on the underlying LLM and their abilities. 
For the very specific case presented above, a simple fix could be to add a constant component that stands last (just before call to action and after best option) that states something like "{name} always picks the best choice or strategy available to it".

**Comment by cth127** [comment]
I understad the mistake (or error) in the final action is also one component of multi-agent system research. However, when examining the log, I find it nearly randomized selection.

How about training a separate 'sample_choice' model which takes CoT as an jnput and categorize it to option? It would potentially result in more robust environment.

**Comment by yhg8423** [comment]
Hi, I experienced a similar issue as well. I've tried adding several components to resolve it, but as @cth127 mentioned, it seems to be more of a random selection. Thus, I wonder if this could be fixed by adding components or others.

**Comment by cth127** [comment]
@jzleibo @vezhnick 

I recommend you consider utilizing `n` argument in `OpenAI` or `TogehterAI` API.
The strategy is simple: sample n times, and choose the frequentist one.
I tested this method on `pub_coordination_mini` 5 times, and no case out of 10 cases (focal agent + background agent) returns the wrong sample choice.
(I know it's a small number, but I observed 6 error cases out of 10 cases for the same agents and seed settings before.)
Here is the code.

```python
  @override
  def sample_choice(
      self,
      prompt: str,
      responses: Sequence[str],
      *,
      seed: int | None = None,
  ) -> tuple[int, str, dict[str, float]]:
      prompt = (
          prompt
          + '\nRespond EXACTLY with one of the following strings:\n'
          + '\n'.join([f'({i})' for i in responses]) + '.'
      )

      messages = [
            {'role': 'user',
             'content': ('You always continue sentences provided ' +
                         'by the user and you never repeat what ' +
                         'the user already said.\n' +
                         'Question: Is Jake a turtle?\nAnswer: Jake is ')},
            {'role': 'assistant',
             'content': 'not a turtle.'},
            {'role': 'user',
             'content': ('Question: What is Priya doing right now?\nAnswer: ' +
                         'Priya is currently ')},
            {'role': 'assistant',
             'content': 'sleeping.'},
            {'role': 'user',
             'content': prompt}
        ]

      sample = ''
      answer = ''
      for attempts in range(_MAX_MULTIPLE_CHOICE_ATTEMPTS):
        if attempts > 0:
          if attempts >= _NUM_SILENT_ATTEMPTS:
            print(
                f'Sleeping for {_SECONDS_TO_SLEEP_WHEN_RATE_LIMITED} seconds... '
                + f'attempt: {attempts} / {_MAX_ATTEMPTS}'
            )
          time.sleep(_SECONDS_TO_SLEEP_WHEN_RATE_LIMITED)
        try:
          self._client = together.Together(api_key=self._api_key, base_url=f"http://192.168.10.2:1000{int(self._count % 3)}/v1")
          response = _client.chat.completions.create(
              model=self._model_name,
              messages=messages,
              max_tokens=3,
              n=20,
              stream=False
          )
        except together.error.RateLimitError as err:
          if attempts >= _NUM_SILENT_ATTEMPTS:
            print(f'  Exception: {err}')
          continue
        else:
          response = [i.message.content for i in response.choices]
          answer_cand = [sampling.extract_choice_response(i) for i in response]
          answer_count = Counter(answer_cand)
          for i in answer_count.keys():
            if i in responses:
              answer=i
              break
          try:
              idx = responses.index(answer)
          except ValueError:
              continue
          else:
              if self._measurements is not None:
                  self._measurements.publish_datum(
                      self._channel, {'choices_calls': attempts}
                  )
              debug = {}
              return idx, responses[idx], debug

      raise language_model.InvalidResponseError(
          (f'Too many multiple choice attempts.\nLast attempt: {sample}, ' +
            f'extracted: {answer}')
      )
```

**Comment by jzleibo** [comment]
I believe you are proposing a change to our language_model API to expose more parameters for choice sampling, which might help sample more consistently. And at least for openai that looks like it would help, since they don't allow logits. For together.ai, since we have logits there, I wonder if the same effect could be achieved by playing with the temperature instead?

Also, I think there's a difference between what the concordia library supports for general research purposes and what we can support specifically as part of the NeurIPS contest. The contest has other constraints like fairness which make it unlikely that a "leap into the unknown" API change at this particular moment would be a good idea.. Of course, for general research purposes, feel free to modify it however you like.

**Comment by cth127** [comment]
> For together.ai, since we have logits there, I wonder if the same effect could be achieved by playing with the temperature instead?

Actually, "6 error cases out of 10 cases" were from the together.ai - logit-based method. 

> Also, I think there's a difference between what the concordia library supports for general research purposes and what we can support specifically as part of the NeurIPS contest. The contest has other constraints like fairness which make it unlikely that a "leap into the unknown" API change at this particular moment would be a good idea.. Of course, for general research purposes, feel free to modify it however you like.

I understand your position. I just wanted to suggest one potential alternative to make an evaluation process as intended (by both organizers and participants).

**Comment by cth127** [comment]
I tested the a33db59511bde861e6531544476ad4a3debbac8f . The result became more consistent, but I still observed many mis-selections.

I tested another alternative method using `echo` argument. Using this we can obtain the logprob of a token in the prompt, not depending on the logprob of newly generated text. Here is the snippet.

```python
    def _sample_choice(response: str) -> float:
      original_prompt = prompt
      prompt = _ensure_prompt_not_too_long(prompt, 1)
      messages = [
          {
              'role': 'user',
              'content': (
                  'You always continue sentences provided '
                  + 'by the user and you never repeat what '
                  + 'the user already said.'
                  +  'Question: Is Jake a turtle?\nAnswer: Jake is '
              ),
          },
          {'role': 'assistant', 'content': 'not a turtle.'},
          {
              'role': 'user',
              'content': (
                  'Question: What is Priya doing right now?\nAnswer: '
                  + 'Priya is currently '
              ),
          },
          {'role': 'assistant', 'content': 'sleeping.'},
          {'role': 'user', 'content': prompt},
          {'role': 'assistant', 'content': response}
      ]

      result = None
      for attempts in range(_MAX_ATTEMPTS):
        if attempts > 0:
          seconds_to_sleep = (_SECONDS_TO_SLEEP_WHEN_RATE_LIMITED +
                              random.uniform(-_JITTER_SECONDS, _JITTER_SECONDS))
          if attempts >= _NUM_SILENT_ATTEMPTS:
            print(
                f'Sleeping for {seconds_to_sleep} seconds.. '
                + f'attempt: {attempts} / {_MAX_ATTEMPTS}'
            )
          time.sleep(seconds_to_sleep)
        try:
          _client = together.Together(api_key=self._api_key)
          result = _client.chat.completions.create(
              model=self._model_name,
              messages=messages,
              max_tokens=1,
              seed=seed,
              logprobs=1,
              stream=False,
              echo=True,
          )
          self._count += 1
        except (together.error.RateLimitError,
                together.error.APIError,
                together.error.ServiceUnavailableError) as err:
          if attempts >= _NUM_SILENT_ATTEMPTS:
            print(f'  Exception: {err}')
            print(f'  Choice exception prompt: {augmented_prompt}')
          if isinstance(err, together.error.APIError):
            # If hit the error that arises from a prompt that is too long then
            # re-run the trimming function with a more pessimistic guess of the
            # the number of characters per token.
            augmented_prompt = _ensure_prompt_not_too_long(
                original_augmented_prompt, 1, guess_chars_per_token=1)
          continue
        else:
          break

      if result:
        logprobs = result.prompt_logprobs
        logprobs.reverse()
        logprobs = [list(i.values())[0] for i in logprobs if i is not None]
        lp = None
        for item in logprobs:
          if item['decoded_token'] == response:
            lp = item['logprob']
            break
      else:
        raise ValueError(
            f'Failed to get logprobs.\nException prompt: {augmented_prompt}')

      return lp
```

**Comment by AlekseyKorshuk** [comment]
+1 
Foud the following issue and added debug prints:
```python
print(f"\n###\n{prompt}\n->{response}\n->{result.choices[0].message.content}\n->{result.choices[0].logprobs}")
```

```
Role playing instructions: The instructions for how to play the role of Alice Lewis are as follows. This is a social science experiment studying how well you play the role of a character named Alice Lewis. The experiment is structured as a tabletop roleplaying game (like dungeons and dragons). However, in this case it is a serious social science experiment and simulation. The goal is to be realistic. It is important to play the role of a person like Alice Lewis as accurately as possible, i.e., by responding in ways that you think it is likely a person like Alice Lewis would respond, and taking into account all information about Alice Lewis that you have. Always use third-person limited perspective.

Overarching goal: Have a good time. To have a good time, Alice Lewis would like to watch the game in the same pub as Alice Lewis, Jack Taylor. Alice Lewis would prefer everyone went to The Golden Fleece.

Current time:  14 May 2015 [18:00 - 18:05]

Observation: [14 May 2015 18:00:00] [observation] Alice Lewis realises it is time to go watch the game at a pub. 
[14 May 2015 18:00:00] [observation] 

Summary of recent observations: [13 May 2015 18:00:00  - 14 May 2015  18:00:00]:  Alice Lewis is planning to watch the Belgium vs. Lithuania game at a pub with her friend Jack Taylor. They are deciding between The Golden Fleece and The Red Lion. 



Recalled memories and observations: [14 May 2015 10:00:00] [observation] Alice Lewis -- "So, about tonight... anyone up for The Golden Fleece? They've got a great atmosphere for games.
[14 May 2015 10:00:00] [observation] Tonight is the night of the game between Belgium and Lithuania. Friends are going to watch the game at a pub, but they are not sure which pub to go to.
[14 May 2015 10:00:00] [observation] Jack Taylor -- "The Red Lion's always good for a game.  Good atmosphere, good beer...
[14 May 2015 18:00:00] [observation] Alice Lewis realises it is time to go watch the game at a pub. 


Question: Which options are available to Alice Lewis right now?
Answer: Based on the information provided, here are some actions available to Alice Lewis:

Question: Of the options available to Alice Lewis, and given their goal, which choice of action or strategy is best for Alice Lewis to take right now?
Answer: Alice Lewis's best course of action is to suggest to Jack Taylor that they go to The Golden Fleece to watch the game. 


Question: To which pub would Alice Lewis go to watch the game?
  (a) The Golden Fleece
  (b) The Red Lion
Answer: (
->b
->(
->tokens=['('] token_logprobs=[-0.76171875] token_ids=[235278]
```

So the issue is that "user" message ends with "Answer: (b" and "Answer: (a" for 2 threads in a single `sample_choice`. Therefore, with `max_tokens=1` and `logprobs=1` we gonna observe anything but the letter of the selection. 
Can we fix this? Otherwise, all the results are random

cc @jzleibo

**Comment by AlekseyKorshuk** [comment]
Created a Pull Request with the fix and provided an example test to validate the result: https://github.com/google-deepmind/concordia/pull/99

**Comment by vezhnick** [comment]
Thanks, looking at it right now

**Comment by cth127** [comment]
@vezhnick After the commit 281847fedc5903c3138b8a9231bc5a1d3242ff21, I still observe inconsistency.
After examining the logprob arguments, I guess just summing whole logprobs does not return consistent result. for example,


- ... {'1261': {'logprob': -4.135303974151611, 'rank': 2, 'decoded_token': 'Answer'}}, {'235292': {'logprob': -0.00023409964342135936, 'rank': 1, 'decoded_token': ':'}}, {'591': {'logprob': -6.276963710784912, 'rank': 9, 'decoded_token': ' ('}}, {'107': {'logprob': -17.865737915039062, 'rank': 738, 'decoded_token': ''}}, {'108': {'logprob': -1.833747386932373, 'rank': 2, 'decoded_token': '\n'}}, {'106': {'logprob': -22.841524124145508, 'rank': 30780, 'decoded_token': ''}}, {'2516': {'logprob': -19.788537979125977, 'rank': 1230, 'decoded_token': 'model'}}, {'108': {'logprob': -1.163489818572998, 'rank': 2, 'decoded_token': '\n'}}, **{'235250': {'logprob': -3.5318799018859863, 'rank': 3, 'decoded_token': 'a'}}, {'107': {'logprob': -24.80474281311035, 'rank': 3322, 'decoded_token': ''}}**, {'108': {'logprob': -2.3021979331970215, 'rank': 3, 'decoded_token': '\n'}}, {'106': {'logprob': -23.668498992919922, 'rank': 20477, 'decoded_token': ''}}, {'2516': {'logprob': -19.03221321105957, 'rank': 1301, 'decoded_token': 'model'}}, {'108': {'logprob': -1.2637945413589478, 'rank': 1, 'decoded_token': '\n'}}

- ... {'1261': {'logprob': -4.135303974151611, 'rank': 2, 'decoded_token': 'Answer'}}, '235292': {'logprob': -0.00023409964342135936, 'rank': 1, 'decoded_token': ':'}}, {'591': {'logprob': -6.276963710784912, 'rank': 9, 'decoded_token': ' ('}}, {'107': {'logprob': -17.865737915039062, 'rank': 738, 'decoded_token': ''}}, {'108': {'logprob': -1.833747386932373, 'rank': 2, 'decoded_token': '\n'}}, {'106': {'logprob': -22.841524124145508, 'rank': 30780, 'decoded_token': ''}}, {'2516': {'logprob': -19.788537979125977, 'rank': 1230, 'decoded_token': 'model'}}, {'108': {'logprob': -1.163489818572998, 'rank': 2, 'decoded_token': '\n'}}, **{'235268': {'logprob': -0.13523344695568085, 'rank': 1, 'decoded_token': 'b'}}, {'107': {'logprob': -30.187583923339844, 'rank': 28399, 'decoded_token': ''}}**, {'108': {'logprob': -2.6002910137176514, 'rank': 3, 'decoded_token': '\n'}}, {'106': {'logprob': -24.384578704833984, 'rank': 31123, 'decoded_token': ''}}, {'2516': {'logprob': -21.940980911254883, 'rank': 2472, 'decoded_token': 'model'}}, {'108': {'logprob': -2.4266862869262695, 'rank': 3, 'decoded_token': '\n'}}


These are the latter part of logprobs for the input prompt, and the bold character represents the `response` part.
I observed that the model first assigns a higher logprob to the consistent token (`b` with 3.4 gap) but later it assigns a much lower logprob to the next token of the `b` (with -6 gap which offsets the previous gap). As a result, the current code chooses `a` as a final choice.

This is not an intuitive result as the token `b` with a much lower logprob is ignored... It would be great if we could extract the logprob of the exact `response` token. (I previously suggested one method but I think it can be broken when the `response` consists of multiple tokens.)

**Comment by vezhnick** [comment]
Thanks, gonna investigate now...

**Comment by vezhnick** [comment]
We are about to submit the change where only the response tokens are taken into account, cutting off all the rest. If you use InteractiveDocument class for sampling choices, this would mean that only single token letter would be evaluated for log_probs.
For example:

prompt = interactive_document.InteractiveDocument(model)
prompt.statement('Alice loves the Princess of Wales.')
answer = prompt.multiple_choice_question('Which pub would Alice got to?', ['the Princess of Wales', 'the Royal Oak','the Squat'])

The tokens and their log_probs will be:
['c']
[-11.625]
['b']
[-0.00064086914]
['a']
[-10.5]

**Comment by cth127** [comment]
That's great! Thanks for the fast response!
